classes/image_type_tezi_run.bbclass

```diff
@@ -150,74 +154,53 @@ def rootfs_tezi_run_create_json(d, flash_type, type_specific_name = False):
     if type_specific_name:
         imagefile = 'image-{0}.json'.format(flash_type)
 
-    with open(os.path.join(deploydir, imagefile), 'w') as outfile:
-         json.dump(data, outfile, indent=4)
-    d.appendVar('TEZI_IMAGE_FILES', imagefile + ' ')
     bb.note("Toradex Easy Installer metadata file {0} written.".format(imagefile))
+    with open(os.path.join(d.getVar('WORKDIR', True), 'tmp-tezirunimage', imagefile), 'w') as outfile:
+         json.dump(data, outfile, indent=4)
 
 python rootfs_tezi_run_json() {
-    if not bb.utils.contains("IMAGE_FSTYPES", "tezirunimg", True, False, d):
-        return
-
     flash_types = d.getVar('TORADEX_FLASH_TYPE', True)
     if flash_types is None:
         bb.fatal("Toradex flash type not specified")
 
     flash_types_list = flash_types.split()
     for flash_type in flash_types_list:
         rootfs_tezi_run_create_json(d, flash_type, len(flash_types_list) > 1)
-
-    # We end up having the same binary twice in TEZI_UBOOT_BINARIES in case
-    # the recovery mode binary is the same as the one which gets flashed.
-    # This is ok, zip will only add it once below.
-    d.appendVar('TEZI_UBOOT_BINARIES', d.getVar('TEZI_UBOOT_BINARY_RECOVERY', True) + ' ')
-}
-
-do_rootfs[depends] =+ "virtual/bootloader:do_deploy u-boot-distro-boot:do_deploy"
-
-IMAGE_CMD_tezirunimg () {
-	bbnote "Create Toradex Easy Installer FIT image"
 }
 
 build_fitimage () {
-	# We have to execute this step after any image generation so we have
-	# squashfs, kernel etc all in DEPLOY_DIR_IMAGE
-	# We cannot work in IMGDEPLOYDIR since that folder only has the image
-	# files in it (not kernel etc...). We could probably make this a
-	# SSTATETASKS similar to how image.class makes do_image_complete...
-	mkimage -f ${DEPLOY_DIR_IMAGE}/tezi.its ${DEPLOY_DIR_IMAGE}/tezi.itb
-}
-
-build_deploytar () {
-	cd ${DEPLOY_DIR_IMAGE}
-
-	# mkdir fails if existing
-	if [ -e ${IMAGE_NAME} ]; then
-		rm -r ${IMAGE_NAME}/
-	fi
-
-	mkdir ${IMAGE_NAME}/
-	cp -L -R ${SPL_BINARY} ${TEZI_UBOOT_BINARIES} tezi.itb wrapup.sh ${TEZI_IMAGE_FILES} tezi.png ${TEZI_DISTRO_BOOT_SCRIPTS} recovery-linux.sh recovery-windows.bat recovery/ ${IMAGE_NAME}/
-
-	# zip does update if the file exist, explicitly delete before adding files to the archive
-	if [ -e ${IMAGE_NAME}.zip ]; then
-		rm ${IMAGE_NAME}.zip
+	# Copy kernel image
+	cp ${DEPLOY_DIR_IMAGE}/${KERNEL_IMAGETYPE} ${WORKDIR}/tmp-fitimage
+
+	# Copy device tree
+	for dtb in ${KERNEL_DEVICETREE}; do
+		cp ${DEPLOY_DIR_IMAGE}/${KERNEL_IMAGETYPE}-$dtb ${WORKDIR}/tmp-fitimage
+	done
+
+	# Copy rootfs image
+	if [ -e ${IMGDEPLOYDIR}/${IMAGE_BASENAME}-${MACHINE}.squashfs ]; then
+		cp ${IMGDEPLOYDIR}/${IMAGE_BASENAME}-${MACHINE}.squashfs ${WORKDIR}/tmp-fitimage
+	else
+		cp ${DEPLOY_DIR_IMAGE}/${IMAGE_BASENAME}-${MACHINE}.squashfs ${WORKDIR}/tmp-fitimage
```

Comment:
Yes, it is. I can explain the idea here.

In the original implementation, why the artifact could not be covered by sstate is we were adding do_assemble_tezirunimg after do_image_complete, and why we were doing that is because do_assemble_tezirunimg requires kernel images, dtbs, and squashfs in the same directory with its file, to assemble the fitimage, so, to be able to make all the files in the same directories with its file, we copy them into ${WORKDIR}/tmp-fitimage, including the squashfs image, and that file could be untared from sstate (in DEPLOY_DIR_IMAGE) or generated by do_image_squashfs (in IMGDEPLOYDIR).

Reply:
Yes, I agree, this part is really tricky.

Reply:
I see what you are saying, but when we have a `IMAGE_TYPEDEP_tezirunimg` it is somehow guaranteed that the squashfs is in the right place...

Can we maybe somehow enable python with IMAGE_CMD?